Feature: New metaprovider for iTunes music

Ivan12273 opened the pull request:
Context: How I mentioned in this issue: https://github.com/nukeop/nuclear/issues/1005 I added iTunes music API as metaprovider.

Hello, I did some modifies on rest/iTunesPodcast.ts, I changed the name to rest/iTunes.ts and added functions for request music. I also created meta/itunesmusic.ts and Implemented the MetaProvider functions.

I ran the tests, I did not have problems, here is a video of how it works :)

https://user-images.githubusercontent.com/42753018/124782922-411cb700-df0a-11eb-943f-64060c0413d6.mp4

nukeop commented:
Hmm timeout after 20 minutes on mac, which is typically much faster. Not sure if it's related to the PR

Ivan12273 commented:
> Hmm timeout after 20 minutes on mac, which is typically much faster. Not sure if it's related to the PR

I see, I am trying to figure out why, I only modified files from core folder so I do not understand what happened.

nukeop commented:
Maybe Github randomly gave us a weaker machine for this run. It's fine.

nukeop commented on packages/core/src/plugins/meta/itunesmusic.ts:
```diff
@@ -0,0 +1,152 @@
+      .filter(artist => artist.artistType === 'Artist')
+      .map(artist => ({
+        id: artist.artistId,
+        coverImage: '',
```
Too bad they don't have images for artists 😔

Ivan12273 replied:
Yeah, I tried but it would require a lot of API request :/

nukeop commented:
Thanks for the contribution. I reviewed the code, it's pretty great. Could you please write some tests to cover your changes? Either unit tests for the provider or a larger integration test for the search box with this provider being selected.

Ivan12273 commented:
> Thanks for the contribution. I reviewed the code, it's pretty great. Could you please write some tests to cover your changes? Either unit tests for the provider or a larger integration test for the search box with this provider being selected.

Alright, is there any reference that I could use? :)

HaiDang666 commented:
> > Thanks for the contribution. I reviewed the code, it's pretty great. Could you please write some tests to cover your changes? Either unit tests for the provider or a larger integration test for the search box with this provider being selected.
> 
> Alright, is there any reference that I could use? :)
one of those file is unit test you can reference.
https://github.com/nukeop/nuclear/blob/master/packages/core/src/rest/audius.test.ts

nukeop commented:
I'll make sure to write some guidelines for creating new tests soon.

Ivan12273 commented:
@nukeop Hi, I just added unit tests for rest/iTunes.ts, meta/itunesmusic.ts and meta/itunespodcast.ts.
I also created mocks and added them to a new folder: core/src/__mocks__

nukeop commented:
I noticed that you put your mocks in a `__mocks__` folder, which has a special function in Jest - it's used for manual mocks that automatically replace corresponding modules: https://jestjs.io/docs/manual-mocks
This shouldn't break anything, but it breaks this convention in a confusing way - can you move them to be adjacent to tests? Or in a folder with a different name.

Ivan12273 commented:
> I noticed that you put your mocks in a `__mocks__` folder, which has a special function in Jest - it's used for manual mocks that automatically replace corresponding modules: https://jestjs.io/docs/manual-mocks
> This shouldn't break anything, but it breaks this convention in a confusing way - can you move them to be adjacent to tests? Or in a folder with a different name.



> I noticed that you put your mocks in a `__mocks__` folder, which has a special function in Jest - it's used for manual mocks that automatically replace corresponding modules: https://jestjs.io/docs/manual-mocks
> This shouldn't break anything, but it breaks this convention in a confusing way - can you move them to be adjacent to tests? Or in a folder with a different name.

Ready, I moved it and changed the name, it is fine?

nukeop commented:
Looks great, thank you.
